src/FsAutoComplete/CodeFixes.fs

```diff
@@ -0,0 +1,719 @@
+                         NewText = replaced } |] } ]
+        | _ -> return []
+      }
+      |> AsyncResult.foldResult id (fun _ -> [])
+
+  /// a codefix that adds in missing '=' characters in type declarations
+  let addMissingEqualsToTypeDefinition (getFileLines: string -> Result<string [], _>) =
+    ifDiagnosticByCode
+      (fun diagnostic codeActionParams ->
+        asyncResult {
+          if diagnostic.Message.Contains "Unexpected symbol '{' in type definition"
+             || diagnostic.Message.Contains "Unexpected keyword 'member' in type definition" then
+            let fileName =
+              codeActionParams.TextDocument.GetFilePath()
+
+            let! lines = getFileLines fileName
+
+            match walkBackUntilCondition lines (dec lines diagnostic.Range.Start) (System.Char.IsWhiteSpace >> not) with
+            | Some firstNonWhitespaceChar ->
+                let insertPos = inc lines firstNonWhitespaceChar
+
+                return
+                  [ { SourceDiagnostic = Some diagnostic
+                      Title = "Add missing '=' to type definition"
+                      File = codeActionParams.TextDocument
+                      Edits =
+                        [| { Range = { Start = insertPos; End = insertPos }
+                             NewText = " =" } |] } ]
+            | None -> return []
+          else
+            return []
+        }
+        |> AsyncResult.foldResult id (fun _ -> []))
+      (Set.ofList [ "10"; "3360" ])
+
+  /// a codefix that corrects -<something> to - <something> when negation is not intended
+  let changeNegationToSubtraction (getFileLines: string -> Result<string [], _>): CodeFix =
+    ifDiagnosticByCode
+      (fun diagnostic codeActionParams ->
+        asyncResult {
+          let fileName =
+            codeActionParams.TextDocument.GetFilePath()
+
+          let! lines = getFileLines fileName
+
+          match walkForwardUntilCondition lines (inc lines diagnostic.Range.End) (fun ch -> ch = '-') with
+          | Some dash ->
+              return
+                [ { SourceDiagnostic = Some diagnostic
+                    Title = "Use subtraction instead of negation"
+                    File = codeActionParams.TextDocument
+                    Edits =
+                      [| { Range = { Start = dash; End = inc lines dash }
+                           NewText = "- " } |] } ]
+          | None -> return []
+        }
+        |> AsyncResult.foldResult id (fun _ -> []))
+      (Set.ofList [ "3" ])
+
+  /// a codefix that corrects == equality to = equality
+  let doubleEqualsToSingleEquality: CodeFix =
+    ifDiagnosticByCode
+      (fun diagnostic codeActionParams ->
+        async.Return [ { Title = "Use '=' for equality check"
+                         File = codeActionParams.TextDocument
+                         SourceDiagnostic = Some diagnostic
+                         Edits =
+                           [| { Range = diagnostic.Range
```

Comment:
sometimes you get really lucky and the diagnostic itself covers the entire range you need to replace.